src/Native/System.Native/pal_networking.cpp

```diff
@@ -624,3 +630,83 @@ SetIPv6Address(uint8_t* socketAddress, int32_t socketAddressLen, uint8_t* addres
 
     return PAL_SUCCESS;
 }
+
+extern "C" int32_t GetControlMessageBufferSize(int32_t isIPv4, int32_t isIPv6)
+{
+    // Note: it is possible that the address family of the socket is neither
+    //       AF_INET nor AF_INET6. In this case both inputs will be false and
+    //       the controll message buffer size should be zero.
+    return (isIPv4 != 0 ? CMSG_SPACE(sizeof(in_pktinfo)) : 0) + (isIPv6 != 0 ? CMSG_SPACE(sizeof(in6_pktinfo)) : 0);
+}
+
+static int32_t GetIPv4PacketInformation(cmsghdr* controlMessage, IPPacketInformation* packetInfo)
+{
+    assert(controlMessage != nullptr);
+    assert(packetInfo != nullptr);
+
+    if (controlMessage->cmsg_len < sizeof(in_pktinfo))
+    {
+        assert(false && "expected a control message large enough to hold an in_pktinfo value");
+        return 0;
+    }
+
+    auto* pktinfo = reinterpret_cast<in_pktinfo*>(CMSG_DATA(controlMessage));
+    ConvertInAddrToByteArray(&packetInfo->Address.Address[0], NUM_BYTES_IN_IPV4_ADDRESS, pktinfo->ipi_addr);
+    packetInfo->InterfaceIndex = static_cast<int32_t>(pktinfo->ipi_ifindex);
+
+    return 1;
+}
+
+static int32_t GetIPv6PacketInformation(cmsghdr* controlMessage, IPPacketInformation* packetInfo)
+{
+    assert(controlMessage != nullptr);
+    assert(packetInfo != nullptr);
+
+    if (controlMessage->cmsg_len < sizeof(in6_pktinfo))
+    {
+        assert(false && "expected a control message large enough to hold an in6_pktinfo value");
+        return 0;
+    }
+
+    auto* pktinfo = reinterpret_cast<in6_pktinfo*>(CMSG_DATA(controlMessage));
```

Comment:
It could just be `auto`, yes. I find `auto*` a bit clearer (esp. since `auto&` is necessary when dealing with references).